### PR TITLE
feat(wallet): smart wallet performAction

### PR DESCRIPTION
**Pull request by samsiegart**
refs: https://github.com/Agoric/agoric-sdk/issues/5914
refs: https://github.com/Agoric/agoric-sdk/issues/5354

### CODE REVIEW

File: packages/wallet/api/src/findOrMakeInvitation.js [hunk 1]
```diff
@@ -94,9 +94,23 @@ const makeContinuingInvitation = async (
   return invitationP;
 };
 
+const makeInvitation = async (
```
Comment (turadg) [inline packages/wallet/api/src/findOrMakeInvitation.js:105]: please add types. `invitationMaker` is a little unconventional.

File: packages/wallet/api/src/findOrMakeInvitation.js [hunk 2]
```diff
@@ -94,9 +94,23 @@ const makeContinuingInvitation = async (
...
+  invitationMaker,
+  instanceHandleBoardId,
+  board,
+  zoe,
+) => {
+  const instance = E(board).getValue(instanceHandleBoardId);
+  const publicFacet = E(zoe).getPublicFacet(instance);
+  const { description, args = [] } = invitationMaker;
```
Comment (turadg) [inline packages/wallet/api/src/findOrMakeInvitation.js]: so `description` specifies a method name? this seems like an abuse of the field. why is this safe?

File: packages/wallet/api/src/lib-wallet.js [hunk 3]
```diff
@@ -1790,6 +1791,29 @@ export function makeWalletRoot({
...
+    const dappOrigin =
+      requestContext.dappOrigin || requestContext.origin || 'unknown';
+    return acceptOffer(`${dappOrigin}#${rawId}`);
+  };
+
+  const handleSuggestIssuerAction = ({ petname, boardId }) =>
+    suggestIssuer(petname, boardId);
+
+  const performAction = async obj => {
+    const { type, data } = JSON.parse(obj.action);
```
Comment (dckc) [inline packages/wallet/api/src/lib-wallet.js:1811]: Coordinating around what goes in `obj.action` could be tricky, but using a `type` descriminator leaves plenty of room for evolution, so I suppose this is fine.

@michaelfig I wonder if you have any thoughts.

File: packages/wallet/api/test/test-lib-wallet.js [hunk 4]
```diff
@@ -1193,6 +1193,166 @@ test('lib-wallet addOffer for autoswap swap', async (/** @type {LibWalletTestCon
   );
 });
 
+test('lib-wallet performAction suggestIssuer', async (/** @type {LibWalletTestContext} */ t) => {
```
Comment (dckc) [inline packages/wallet/api/test/test-lib-wallet.js]: :+1:

File: packages/wallet/api/test/test-lib-wallet.js [hunk 5]
```diff
@@ -1193,6 +1193,166 @@ test('lib-wallet addOffer for autoswap swap', async (/** @type {LibWalletTestCon
...
+
+  assert(done);
+  t.is(
+    wallet.getIssuer('bucksIssuer'),
+    bucksIssuer,
+    `bucksIssuer is stored in wallet`,
+  );
+});
+
+test('lib-wallet performAction acceptOffer', async (/** @type {LibWalletTestContext} */ t) => {
```
Comment (dckc) [inline packages/wallet/api/test/test-lib-wallet.js]: :+1:

**Review by dckc — APPROVED** [review]
thanks for thorough unit tests

File: packages/wallet/api/src/lib-wallet.js [hunk 6]
```diff
@@ -1790,6 +1791,29 @@ export function makeWalletRoot({
...
+    switch (type) {
+      case 'acceptOffer':
+        return handleAcceptOfferAction(data);
+      case 'suggestIssuer':
+        return handleSuggestIssuerAction(data);
+      default:
+        throw new Error(`Unknown wallet action ${type}`);
+    }
+  };
```
Comment (turadg) [inline packages/wallet/api/src/lib-wallet.js]: This pattern requires naming things twice. consider,
```js
const handlers = {
  async acceptOffer(offer) {
  …
  },
  suggestIssuer({ petname, boardId }) {
    return suggestIssuer(petname, boardId);
  }
};
const performAction = obj => {
  const { type, data } = JSON.parse(obj.action);
  const handle = handlers[type];
  if (!handle) {
    throw new Error(`Unknown wallet action ${type}`);
  }
  return handle(data);
```

Also note `performAction` need not by synchronous. It does need to return a promise. Giving it jsdoc can make that clear.

File: packages/wallet/api/test/test-lib-wallet.js (hunk 4 above)
Comment (turadg) [inline packages/wallet/api/test/test-lib-wallet.js]: context should be typed on `test` binding so they don't have to be repeated in each call. example: https://github.com/Agoric/agoric-sdk/blob/936f9af005eab2333f6460eaeb79a696c60646ce/packages/inter-protocol/test/test-gov-collateral.js#L35-L37

File: packages/wallet/api/test/test-lib-wallet.js [hunk 7]
```diff
@@ -1193,6 +1193,166 @@ test('lib-wallet addOffer for autoswap swap', async (/** @type {LibWalletTestCon
...
+  const { issuer: bucksIssuer } = makeIssuerKit('bucks');
+  const bucksIssuerBoardId = await E(board).getId(bucksIssuer);
+
+  const action = JSON.stringify({
+    type: 'suggestIssuer',
+    data: { petname: 'bucksIssuer', boardId: bucksIssuerBoardId },
+  });
+  const done = await wallet.performAction({ action });
+
+  assert(done);
```
Comment (turadg) [inline packages/wallet/api/test/test-lib-wallet.js]: failure message would be more informative with:
```suggestion
  t.truthy(done);
```

File: packages/wallet/api/test/test-lib-wallet.js [hunk 8]
```diff
@@ -1193,6 +1193,166 @@ test('lib-wallet addOffer for autoswap swap', async (/** @type {LibWalletTestCon
   );
 });
 
+test('lib-wallet performAction suggestIssuer', async (/** @type {LibWalletTestContext} */ t) => {
+  t.plan(1);
```
Comment (turadg) [inline packages/wallet/api/test/test-lib-wallet.js]: planning is helpful when the checks run in control structures. given the checks are all at top level of this function, this seems superfluous and means another thing to change when changing the test.

File: packages/wallet/api/src/findOrMakeInvitation.js (hunk 2 above)
Comment (samsiegart) [reply packages/wallet/api/src/findOrMakeInvitation.js]: I'm not sure what you mean by safe, is `description` misleading of a name? Maybe `descriptor` or `method` instead?

The user has to approve this, so I don't see it as being less safe than the dapp passing in an arbitrary invitation like the other way.

Comment (turadg) [reply packages/wallet/api/src/findOrMakeInvitation.js]: yeah, `method` would address my concern. `description` sounds like a general field for arbitrary text but the string in this field has powers, specifically to specify the method to call.

**Comment by samsiegart** [comment]
> thanks for thorough unit tests

I didn't have much of a choice, it really is the fastest way to test changes 😄

File: packages/wallet/api/src/findOrMakeInvitation.js (hunk 1 above)
Comment (samsiegart) [reply packages/wallet/api/src/findOrMakeInvitation.js:105]: Done

File: packages/wallet/api/src/findOrMakeInvitation.js (hunk 2 above)
Comment (samsiegart) [reply packages/wallet/api/src/findOrMakeInvitation.js]: Changed to method

File: packages/wallet/api/src/lib-wallet.js (hunk 6 above)
Comment (samsiegart) [reply packages/wallet/api/src/lib-wallet.js]: Added a jsdoc.

I think for the `handlers` thing, I'd rather flatten it out into separate functions like it is now, I think it's slightly easier to understand which actions are defined this way.

File: packages/wallet/api/test/test-lib-wallet.js (hunk 4 above)
Comment (samsiegart) [reply packages/wallet/api/test/test-lib-wallet.js]: Done

File: packages/wallet/api/test/test-lib-wallet.js (hunk 8 above)
Comment (samsiegart) [reply packages/wallet/api/test/test-lib-wallet.js]: Removed

File: packages/wallet/api/test/test-lib-wallet.js (hunk 7 above)
Comment (samsiegart) [reply packages/wallet/api/test/test-lib-wallet.js]: Done

**Pull request merged**